### PR TITLE
docs: record ADR-0003 for the admin dashboard extension and cover the combined router surface

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Groundwork for extending shunt's admin surface, prompted by two reference CLIProxyAPI panels (CPA Manager Plus, CPA Usage Keeper). This PR contains **no production source changes** — it is documentation plus tests.

- **ADR-0003 (`.please/docs/decisions/0003-admin-dashboard-extension.md`)** records three decisions. The admin UI platform work — the planned `/admin/api/*` split and an embedded React/Vite SPA — lands before new data features. Request/usage history, when it lands, is backed by optional SQLite, and the in-memory path stays the default. Monetary cost estimates are deferred, because shunt's primary case is consumer subscriptions with no per-token bill. The ADR rests on a survey finding that shunt today retains no per-request event history and computes no cost anywhere, and that `[server.spend]` stores limit policy and audit rather than consumption.
- **The design-record correction (`docs/admin-ui-delivery.md`)** fixes the blocked-path table that Resolution 6 designates as the migration inventory. The table omitted `POST /admin/accounts/claude/{name}/refresh`, so it listed 12 of 13 paths, and the "Current surface" row credited `admin_router` with M9's endpoint-table counts (15/17) rather than the router's own (16/18).
- **The new test file (`tests/router_surface.rs`)** closes the gap recorded under "Risks" in that design record. `axum::Router::merge` panics on a duplicate path+method, but only when both trees are registered, and no test built a router with every optional surface enabled at once. Six tests cover it: the all-surfaces build, a runtime path-inventory probe that uses the 404-vs-405 distinction as a path-existence oracle, an undocumented-path check, a `HEAD /` regression, and a compile-time `include_str!` source scan that catches a route added at a novel path — which the runtime probes alone do not.

## Milestone / spec

- ADR-0003 — `.please/docs/decisions/0003-admin-dashboard-extension.md`
- `docs/admin-ui-delivery.md` (blocked-path inventory for Resolution 6; "Risks" section)

## Checklist

- [x] `cargo build` passes
- [x] `cargo test` passes (new behavior is covered; tests run without network/loopback where possible)
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] Source files stay under 500 lines
- [x] English only; matches surrounding style
- [x] Frozen spec in `docs/` updated if this change deviates from it
- [x] User-facing docs updated for behavior/config/endpoint/CLI/provider/model changes — `README.md` / `site/` as applicable (`wiki/` is generated; don't hand-edit) — n/a, no user-facing behavior changes
- [x] Any new GitHub Action is pinned to a full commit SHA — n/a, no workflow changes

## Verification

- `cargo fmt --all --check` → exit 0
- `cargo clippy --all-targets --all-features -- -D warnings` → exit 0, zero warning/error lines
- `cargo test --all-features --workspace` → exit 0, 2,547 passed, 0 failed
- Each new test was additionally proven non-vacuous by mutation: deleting a route, adding an undocumented route, duplicating `GET /v1/models` across two surfaces, and adding a route at a novel path each produced the expected distinct failure.

## Notes for reviewers

- No production source changes — review scope is the ADR's reasoning, the corrected path inventory, and whether the router-surface tests fail for the right reasons.
- The path-existence oracle in `tests/router_surface.rs` relies on axum returning 405 (not 404) for a registered path with an unmatched method; the compile-time source scan exists because that oracle cannot see a route registered at a path nobody probes.

---
## Summary by cubic
Adds ADR-0003 and a router-surface test suite to establish the admin dashboard baseline and catch route conflicts before UI work begins. No production source changes.

**Documentation**
- Records the React/Vite SPA and `/admin/api/*` split as the first track, optional SQLite history, and deferred cost estimates.
- Corrects the admin inventory, including the Claude refresh and spend-limit routes, current counts, source location, asset-size date, and spend namespace details.
- Marks the duplicate-route risk as covered by the all-surfaces router smoke test.

**Tests**
- Builds the router with every optional surface enabled and checks documented and undocumented paths.
- Pins each path's method set through the `Allow` header, including `HEAD` behavior.
- Scans literal and indirectly registered paths, and pins the `.merge(`/`.nest(` composition points, the non-literal `.route(` call count, and the `.route_service(`/`.nest_service(` counts (at zero), so every registration form must be added to the reviewed inventory.

<sup>Written for commit e4b93197ca572133b0d0c5f8eea21b9371655f21. Summary will update on new commits.</sup>

### CODE REVIEW

File: tests/router_surface.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+fn all_surfaces_config(label: &str) -> Config {
+    let suffix = format!("{}_{label}", std::process::id());
+    let admin_env = format!("SHUNT_ROUTER_SURFACE_ADMIN_{suffix}");
+    let client_env = format!("SHUNT_ROUTER_SURFACE_CLIENT_{suffix}");
+    let jwt_env = format!("SHUNT_ROUTER_SURFACE_JWT_{suffix}");
+    let users_env = format!("SHUNT_ROUTER_SURFACE_USERS_{suffix}");
+    std::env::set_var(&admin_env, "admin:admin-secret");
+    std::env::set_var(&client_env, "tester:client-secret");
+    std::env::set_var(&jwt_env, "0123456789abcdef0123456789abcdef");
+    std::env::set_var(&users_env, "dev@example.com:password");
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:194]: Declining this one, with reasons.

`tokens_env`, `jwt_secret_env`, and `users_env` are public config keys whose *value is the name of an environment variable*. Resolution reads the process environment by design (`src/config.rs:367`, `:508`, `:747`, `:754`) — that is shunt's documented credential-loading contract, not incidental coupling in the test. Moving that behind env-free helpers would change a public config surface, which `AGENTS.md` gates behind an explicit ask, and it is well outside the scope of a router-surface smoke test.

On the flakiness concern specifically: the env var names already carry a `std::process::id()`-derived suffix (`all_surfaces_config`), so no other test in the binary can observe or satisfy them. The comment also describes "setting and resetting" — the test never resets them, deliberately: resetting a shared global at test boundaries is what breaks neighbouring tests, not what protects them.

The same pattern (unique suffix + `set_var`) is the established convention here — see `src/config.rs:4471-4473`.

File: tests/router_surface.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    "/.well-known/oauth-authorization-server",
+    "/oauth/device_authorization",
+    "/oauth/token",
+    "/device",
+    "/device/authorize",
+    "/device/callback",
+    "/managed/settings",
+    "/v1/metrics",
+    "/v1/logs",
+    "/v1/traces",
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Confirmed and fixed in fcebba64. `[server.spend]` does register `/v1/organizations/spend_limits` and `/v1/organizations/spend_limits/{id}` (`src/gateway/spend/mod.rs:20-32`, merged at `src/server.rs:245`), so the "Current surface" table — which calls itself the conflict baseline for any new route — was missing a gate, and the reserved-namespace section's "No shunt-owned route ever claims `/v1/organizations/*`" was unconditionally false. Both rows are added and the invariant is now scoped to the paths outside that two-path subset, with `/effective`, `/audit`, and the metering behind them called out as still unimplemented.

File: tests/router_surface.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+/// no route table, so a route added at a path this file has never heard of
+/// would answer requests without failing any probe. Reading the registrations
+/// back out of the source closes that, which is what makes the inventory an
+/// actual gate rather than a spot check.
+///
+/// Residual gap, stated rather than hidden: a route whose path is not a string
+/// literal at the call site is invisible here. There are two such sites today —
+/// the OTLP signals in `gateway_router` and the `codex_endpoint::PATHS` /
+/// `codex_analytics::PATHS` loops in `build_router` — and both are pinned by
+/// `every_documented_path_is_registered_when_all_surfaces_are_enabled`.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Correct, and fixed in fcebba64. I reproduced it before changing anything: appending `/v1/responses/beta` to `codex_endpoint::PATHS` left all six pre-existing tests green. The file also claimed otherwise in two places — the `CODEX_ENDPOINT_PATHS` mirror said "a change to either constant must fail this test", and the scan said both indirect sites were "pinned by" the documented-path test. That test only pins the removal/rename direction, never an addition. `INDIRECT_PATH_SOURCES` now scans `codex_endpoint.rs`, `codex_analytics.rs`, and `telemetry_ingest.rs` (`Signal::path()`) and compares them against the inventory, with a count assertion so the scan cannot go vacuous. Same mutation now fails both new tests; renaming an OTLP signal path fails the inventory check too. Both comments were rewritten to state the asymmetry rather than overclaim.

File: .please/docs/decisions/0003-admin-dashboard-extension.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+`upstream_status.rs`, and the `gateway/` and `gateway/spend/` trees — establishes
+three facts:
+
+1. **There is no per-request event history anywhere.** shunt retains live
+   aggregates (Sentry/OTLP counters and histograms), per-stream diagnostics that
+   are discarded when the stream finishes (`src/stream_metrics.rs:142-199`), and
+   current per-account quota snapshots. `[server.pool] state_path` persists only
+   the current `QuotaState` (`src/state_persist.rs:52-64`).
+2. **Nothing computes monetary cost.** No pricing table, no currency field, on
+   any path.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/decisions/0003-admin-dashboard-extension.md]: Confirmed and fixed in fcebba64. `SpendLimit.currency` (`src/gateway/spend/store.rs:54`) and the optional request field (`src/gateway/spend/api.rs:43`) both exist, both USD-only, so "no currency field, on any path" was wrong. The ADR now separates that policy field from the absent consumption accounting and keeps the narrower claim that survives: no pricing table anywhere, and nothing multiplies usage by a rate. The decision to defer cost estimates is unchanged; only the baseline it rests on is corrected.

File: tests/router_surface.rs [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+const SPEND_PATHS: [&str; 2] = [
+    "/v1/organizations/spend_limits",
+    "/v1/organizations/spend_limits/{id}",
+];
+
+/// Mirrors `codex_endpoint::PATHS` and `codex_analytics::PATHS`, which are
+/// `pub(crate)` and so cannot be imported here. Duplicating them is deliberate:
+/// a change to either constant must fail this test and be re-reviewed against
+/// the path split, which is exactly the guard being installed.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Correct — fixed in fcebba64, and independently reported by Greptile on the same lines. Verified by mutation before fixing: a fourth path appended to `codex_endpoint::PATHS` registered a live route with all six tests green. The inventory now compares against the constants' own definitions rather than only mirroring their current values, so an addition fails.

File: tests/router_surface.rs [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    "/device",
+    "/device/authorize",
+    "/device/callback",
+    "/managed/settings",
+    "/v1/metrics",
+    "/v1/logs",
+    "/v1/traces",
+];
+
+const SPEND_PATHS: [&str; 2] = [
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Fixed in fcebba64 — same finding Greptile raised. Both spend rows are in the "Current surface" table and the reserved-namespace wording is scoped to the paths outside the `[server.spend]` subset.

File: tests/router_surface.rs [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    "/v1/organizations/spend_limits",
+    "/v1/organizations/spend_limits/{id}",
+];
+
+/// Mirrors `codex_endpoint::PATHS` and `codex_analytics::PATHS`, which are
+/// `pub(crate)` and so cannot be imported here. Duplicating them is deliberate:
+/// a change to either constant must fail this test and be re-reviewed against
+/// the path split, which is exactly the guard being installed.
+const CODEX_ENDPOINT_PATHS: [&str; 5] = [
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Fixed in fcebba64 — same finding Greptile and the Codex reviewer raised. The assertion now scans the defining sources of `codex_endpoint::PATHS`, `codex_analytics::PATHS`, and `Signal::path()`, proven by mutation in both directions.

File: tests/router_surface.rs [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+/// Both `state_path` values are set to an empty path — the documented opt-out
+/// — so building a router never reads or writes the operator's real
+/// `~/.shunt` state.
+fn all_surfaces_config(label: &str) -> Config {
+    let suffix = format!("{}_{label}", std::process::id());
+    let admin_env = format!("SHUNT_ROUTER_SURFACE_ADMIN_{suffix}");
+    let client_env = format!("SHUNT_ROUTER_SURFACE_CLIENT_{suffix}");
+    let jwt_env = format!("SHUNT_ROUTER_SURFACE_JWT_{suffix}");
+    let users_env = format!("SHUNT_ROUTER_SURFACE_USERS_{suffix}");
+    std::env::set_var(&admin_env, "admin:admin-secret");
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:191]: Fixed in fcebba64. I checked the convention claim before acting on it: `tests/admin_surface.rs` is 91 `set_var` to 92 `remove_var`, removing at the end of each test. `all_surfaces_config` now returns an `EnvVars` drop guard that each test binds, so removal happens at end of test body — not on entry, which is the variant that breaks a neighbour mid-run. Agreed this is hygiene rather than isolation, since the pid+label suffix already prevents collisions; the comment says so.

File: .please/docs/decisions/0003-admin-dashboard-extension.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/decisions/0003-admin-dashboard-extension.md]: Fixed in fcebba64 — same finding the Codex reviewer raised. The ADR now distinguishes the USD-only spend-limit policy currency from the missing consumption accounting.

File: tests/router_surface.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:194]: Follow-up for accuracy, since this thread is already closed: fcebba64 does now remove these variables, via a drop guard, after cubic pointed out that `tests/admin_surface.rs` pairs all 91 of its `set_var` calls with a `remove_var`. That does not change the answer above — removal happens on drop at the *end* of the test, never as a reset on entry, and the config keys themselves are untouched — but "the test never resets them" is no longer the whole picture, so I would rather say so here than leave it standing.

File: tests/router_surface.rs [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+/// loops in `build_router`. `every_documented_path_is_registered_when_all_surfaces_are_enabled`
+/// catches a removal or rename in either set but **not** an addition, so
+/// [`INDIRECT_PATH_SOURCES`] scans those definitions to close that direction.
+#[test]
+fn every_registered_literal_path_is_documented() {
+    let inventory = all_registered_paths();
+    for (file, source) in ROUTER_SOURCES {
+        for path in registered_literal_paths(source) {
+            assert!(
+                inventory.contains(&path),
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:481]: Fixed in f8c8befc — you and the other reviewer reached this independently, which is what moved it from "different axis than this gate" to worth doing now.

It turned out to cost nothing extra. The `405` answering the PATCH probe this file already sends carries an `Allow` header naming exactly the registered methods, so the one request that answers "does this path exist" also answers "with which methods", and no handler runs — no state mutated, no auth needed. The six inventory constants are now `(path, allow)` pairs compared as sets; `all_registered_paths` is derived from them, so the four source scans are unchanged.

Verified non-vacuous by mutation: removing `.post(codex::add_codex_account)` from `/admin/accounts/codex` fails the new test with `left: {"GET", "HEAD"}` against `right: {"GET", "HEAD", "POST"}`, while the other eight stay green — the gap you described, reproduced.

Side benefit: counting the `allow` column across the 16 `/admin` paths gives exactly 18 method+path pairs, which independently reproduces the figure `docs/admin-ui-delivery.md` has recorded all along.

File: tests/router_surface.rs [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+}
+
+/// The direction `every_documented_path_is_registered_when_all_surfaces_are_enabled`
+/// cannot cover. That test proves each *documented* path is live, so removing or
+/// renaming a member of one of these sets fails it — but appending a member
+/// leaves it green, because nothing probes a path this file has never heard of.
+/// Reading the definitions back out of their own source closes that direction.
+#[test]
+fn every_indirectly_registered_path_is_documented() {
+    let inventory = all_registered_paths();
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:566]: Fixed in f8c8befc — you and the other reviewer reached this independently, which is what moved it from "different axis than this gate" to worth doing now.

It turned out to cost nothing extra. The `405` answering the PATCH probe this file already sends carries an `Allow` header naming exactly the registered methods, so the one request that answers "does this path exist" also answers "with which methods", and no handler runs — no state mutated, no auth needed. The six inventory constants are now `(path, allow)` pairs compared as sets; `all_registered_paths` is derived from them, so the four source scans are unchanged.

Verified non-vacuous by mutation: removing `.post(codex::add_codex_account)` from `/admin/accounts/codex` fails the new test with `left: {"GET", "HEAD"}` against `right: {"GET", "HEAD", "POST"}`, while the other eight stay green — the gap you described, reproduced.

Side benefit: counting the `allow` column across the 16 `/admin` paths gives exactly 18 method+path pairs, which independently reproduces the figure `docs/admin-ui-delivery.md` has recorded all along.

File: tests/router_surface.rs [hunk 11]
```diff
@@ -0,0 +1,595 @@
...
+    }
+}
+
+/// The gap recorded under "Risks" in `docs/admin-ui-delivery.md`: `Router::merge`
+/// panics on a duplicate path+method, but only when both trees are registered.
+#[tokio::test]
+async fn every_optional_surface_can_be_enabled_at_once() {
+    let (config, _env) = all_surfaces_config("builds");
+    let (_router, _shared, _state) =
+        server::build_router(config).expect("a config enabling every optional surface builds");
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:374]: Fixed in 1664907d. You are right that this PR is what made it stale: `every_optional_surface_can_be_enabled_at_once` builds exactly that router, so "No test builds a router with `[server.admin]`, `[server.gateway]`, and `[server.codex_endpoint]` enabled together" stopped being true the moment this branch landed, while the Testing section already described the test in the future tense.

The Risks bullet now names the test that closes it and keeps the hazard on the page rather than deleting the entry — `Router::merge` only panics when both trees are registered, which is precisely why the smoke test has to keep every surface enabled, so a future reader tempted to narrow it should see the reason.

File: tests/router_surface.rs [hunk 12]
```diff
@@ -0,0 +1,595 @@
...
+/// at compile time, so this test needs no filesystem access at runtime.
+const ROUTER_SOURCES: [(&str, &str); 4] = [
+    ("src/server.rs", include_str!("../src/server.rs")),
+    ("src/admin/mod.rs", include_str!("../src/admin/mod.rs")),
+    ("src/gateway/mod.rs", include_str!("../src/gateway/mod.rs")),
+    (
+        "src/gateway/spend/mod.rs",
+        include_str!("../src/gateway/spend/mod.rs"),
+    ),
+];
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:444]: Fixed in 6701b683. I checked the present state first: today the four scanned files do cover every production registration — the other `.route(` calls under `src/` are all inside `#[cfg(test)]` modules (`src/concurrency.rs:161`, `src/http_tuning.rs:162`), and `liveness_router` is built in `src/server.rs` itself, which is scanned. So this was a future hole rather than a live one.

Rather than try to discover modules, the new test guards the composition point: `build_router` makes exactly four `.merge(` calls and no `.nest(` call, and both are pinned. It does not find a new router — it fails when one is composed in, so the addition gets the path-split review and `ROUTER_SOURCES` is extended in the same change. The comment says that explicitly rather than claiming discovery it does not do. `.nest(` is asserted separately because nesting rewrites the paths its routes answer on, which would invalidate both scans rather than merely widen them.

Mutation-verified: composing a fifth tree into `build_router` fails it with `left: 5` against `right: 4` while the other nine tests stay green.

File: tests/router_surface.rs [hunk 13]
```diff
@@ -0,0 +1,632 @@
...
+fn registered_literal_paths(source: &str) -> Vec<&str> {
+    source
+        .match_indices(".route(")
+        .filter_map(|(index, marker)| {
+            let rest = source[index + marker.len()..].trim_start();
+            // A non-literal first argument (`telemetry_ingest::Signal::…path()`,
+            // or the `path` loop variable in `server.rs`) is out of this scan's
+            // reach by construction; those paths are covered by the runtime
+            // probes above instead.
+            let literal = rest.strip_prefix('"')?;
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:456]: Fixed in 8640c17e. The skip was the hazard, not the non-literal call itself: `registered_literal_paths` dropped those arguments *silently*, so a future `.route(NEW_PATH, ...)` would have changed neither the literal count nor the inventory. The skips are now counted and pinned at the five indirect sites `INDIRECT_PATH_SOURCES` already reads (the two Codex `PATHS` loops in `build_router`, the three `Signal::path()` calls in `gateway_router`).

Mutation-verified: adding `router.route(MUTATION_PATH, ...)` to `build_router` fails it with `left: 6` against `right: 5`, while the other ten tests stay green — the bypass as described.

File: tests/router_surface.rs [hunk 14]
```diff
@@ -0,0 +1,632 @@
...
+/// on a path that already exists changes no path string, so every other test in
+/// this file stays green while the surface has actually moved. The `405` probe
+/// already names the registered methods in its `Allow` header, so pinning them
+/// costs no extra request and runs no handler.
+#[tokio::test]
+async fn every_registered_method_set_matches_the_inventory() {
+    let (config, _env) = all_surfaces_config("methods");
+    let (router, _shared, _state) = server::build_router(config).expect("router builds");
+
+    for (path, documented) in all_registered_entries() {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:355]: Fixed in 8640c17e. The skip was the hazard, not the non-literal call itself: `registered_literal_paths` dropped those arguments *silently*, so a future `.route(NEW_PATH, ...)` would have changed neither the literal count nor the inventory. The skips are now counted and pinned at the five indirect sites `INDIRECT_PATH_SOURCES` already reads (the two Codex `PATHS` loops in `build_router`, the three `Signal::path()` calls in `gateway_router`).

Mutation-verified: adding `router.route(MUTATION_PATH, ...)` to `build_router` fails it with `left: 6` against `right: 5`, while the other ten tests stay green — the bypass as described.

File: tests/router_surface.rs [hunk 15]
```diff
@@ -0,0 +1,632 @@
...
+/// review `docs/admin-ui-delivery.md` asks for and `ROUTER_SOURCES` gets extended
+/// in the same change. A bare `.route(` added directly to `build_router` needs no
+/// guard — `src/server.rs` is itself scanned.
+#[test]
+fn no_router_tree_is_composed_in_from_an_unscanned_module() {
+    let server = ROUTER_SOURCES
+        .iter()
+        .find(|(file, _)| *file == "src/server.rs")
+        .map(|(_, source)| *source)
+        .expect("src/server.rs is one of the scanned router sources");
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Fixed in 8640c17e. You are right that guarding only `build_router` missed the point: `gateway_router` merging a tree from a new module hides it exactly as well. The guard now runs over every `ROUTER_SOURCES` entry, with per-file counts in `COMPOSITION_COUNTS` (4 merges in `src/server.rs`, 0 in the three child modules, 0 nests anywhere).

Mutation-verified: a `.merge(` inside `gateway_router` fails it with `left: 1` against `right: 0`, while the other ten stay green.

With this and the non-literal count, the scanned set now has a statable invariant rather than an implicit boundary: no registration is silently dropped — every `.route(` either resolves to a literal path that must appear in the inventory or is one of the five tracked indirect sites, and every `.merge(`/`.nest(` is counted. Further widening (discovering modules outside the scanned set at all) is a different mechanism and belongs in its own change, not this PR.

File: tests/router_surface.rs [hunk 16]
```diff
@@ -0,0 +1,632 @@
...
+
+    // `admin_router`, `gateway_router`, `spend_router`, and the liveness tree
+    // built in `src/server.rs` itself.
+    assert_eq!(
+        server.matches(".merge(").count(),
+        4,
+        "src/server.rs composes a different number of router trees than the four this test knows \
+         about. If a new one was added, add its source file to ROUTER_SOURCES so its paths are \
+         scanned, and review the routes it brings against the /admin, /admin/api and reserved \
+         /v1/organizations path split in docs/admin-ui-delivery.md."
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs]: Fixed in 8640c17e. You are right that guarding only `build_router` missed the point: `gateway_router` merging a tree from a new module hides it exactly as well. The guard now runs over every `ROUTER_SOURCES` entry, with per-file counts in `COMPOSITION_COUNTS` (4 merges in `src/server.rs`, 0 in the three child modules, 0 nests anywhere).

Mutation-verified: a `.merge(` inside `gateway_router` fails it with `left: 1` against `right: 0`, while the other ten stay green.

With this and the non-literal count, the scanned set now has a statable invariant rather than an implicit boundary: no registration is silently dropped — every `.route(` either resolves to a literal path that must appear in the inventory or is one of the five tracked indirect sites, and every `.merge(`/`.nest(` is counted. Further widening (discovering modules outside the scanned set at all) is a different mechanism and belongs in its own change, not this PR.

File: tests/router_surface.rs [hunk 17]
```diff
@@ -0,0 +1,678 @@
...
+/// Counting the skips closes that. Together with the two count assertions above
+/// and the composition guard, the invariant across the scanned files is that no
+/// registration is silently dropped: every `.route(` either resolves to a literal
+/// path that must appear in the inventory, or is one of these five indirect sites
+/// whose definitions [`INDIRECT_PATH_SOURCES`] reads.
+#[test]
+fn every_nonliteral_route_call_is_one_this_test_already_tracks() {
+    let calls: usize = ROUTER_SOURCES
+        .iter()
+        .map(|(_, source)| source.matches(".route(").count())
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/router_surface.rs:683]: Fixed in e4b93197 — and you caught my own claim, not just a gap. The previous commit stated an invariant ("no registration is silently dropped") that its assertions did not deliver: `.route(` is not a substring of `.route_service(`, so nothing in the file saw axum's service-based registrations.

`.route_service(` and `.nest_service(` are now counted beside `.merge(` and `.nest(`, pinned at zero in every scanned source, and the invariant's wording names the four APIs actually enumerated instead of gesturing at all of them.

`Router::fallback` still gets no count, for a stated reason rather than by omission: it claims no path, and a catch-all would make `path_is_registered` answer something other than `404`/`405` for an unrouted path, which it panics on. The two `.fallback(` calls in `spend_router` are `MethodRouter::fallback` — an unregistered method on a path that does exist — which the `Allow` probe already covers. Conflating those two is what would make the next count assertion wrong, so both are written down.

Mutation-verified: `.route_service("/mutation-svc", get(...))` in `build_router` fails with `left: 1` against `right: 0` while the other ten stay green. Worth noting my first mutation attempt exited 101 as a *compile* error (`tower` not in scope), which would have read as a pass if I had trusted the exit code — the evidence is the panic text.
